### ISSUE TITLE
KeyError in _t_child from weakref

### ISSUE THREAD

**Issue by tdsmith** [mahmoud/glom#48]
I'm running into a weird issue using glom in PySpark on Databricks.

This expression:

`glom(ping, (T[stub]["values"].values(), sum), default=0)`

(where `stub` is `"a11y_time"`)

is consistently throwing this exception when I run it on my real data:

```
/databricks/spark/python/lib/py4j-0.10.6-src.zip/py4j/protocol.py in get_return_value(answer, gateway_client, target_id, name)
     318 raise Py4JJavaError(
     319 "An error occurred while calling {0}{1}{2}.\n". 
 --> 320                     format(target_id, ".", name), value)
     321 else:
     322 raise Py4JError( Py4JJavaError: An error occurred while calling z:org.apache.spark.api.python.PythonRDD.runJob. : org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 22.0 failed 4 times, most recent failure: Lost task 0.3 in stage 22.0 (TID 243, 10.166.248.213, executor 2): org.apache.spark.api.python.PythonException: 

Traceback (most recent call last):
File "/databricks/spark/python/pyspark/worker.py", line 229, in main process()
File "/databricks/spark/python/pyspark/worker.py", line 224, in process serializer.dump_stream(func(split_index, iterator), outfile)
File "/databricks/spark/python/pyspark/serializers.py", line 372, in dump_stream vs = list(itertools.islice(iterator, batch))
File "/databricks/spark/python/pyspark/rdd.py", line 1354, in takeUpToNumLeft yield next(iterator) 
File "<command-26292>", line 10, in to_row
File "<command-26292>", line 5, in histogram_measures
File "/databricks/python/local/lib/python2.7/site-packages/glom/core.py", line 753, in __getitem__ return _t_child(self, '[', item)
File "/databricks/python/local/lib/python2.7/site-packages/glom/core.py", line 791, in _t_child _T_PATHS[t] = _T_PATHS[parent] + (operation, arg)
File "/usr/lib/python2.7/weakref.py", line 330, in __getitem__ return self.data[ref(key)]
KeyError: <weakref at 0x7f84c7d2f6d8; to '_TType' at 0x7f84c8933f30>
```

The object that's crashing it is, itself, totally unremarkable:

`{'submission_date': u'20180718', 'a11y_count': None, 'a11y_node_inspected_count': None, 'a11y_service_time': None, 'toolbox_time': None, 'toolbox_count': None, 'a11y_time': None, 'branch': u'Treatment', 'client_id': u'some-random-uuid', 'a11y_picker_time': None, 'a11y_select_accessible_for_node': None} `

The Python that Databricks is running looks like `2.7.12 (default, Dec 4 2017, 14:50:18) [GCC 5.4.0 20160609]`.

I can't reproduce it on my Mac in 2.7.14 or 2.7.12.

**Comment by mahmoud** [comment]
Yeah that's a pretty straightforward case (covered by the tests, etc.), so it's odd that it'd fail. Some idiosyncrasy around Python and PySpark? Maybe just a latent weakref-related bug that's coming out because your data is bigger and more collections are happening. 

Thanks for the report, and I'll take a closer look shortly!

**Comment by kurtbrose** [comment]
Best guess is pickling.

Because we store the path itself in a global weakref dict the default pickle behavior leaves it behind.

Need to check if we can override the behavior, or maybe rethink the design of having T instances be "empty".

**Comment by tdsmith** [comment]
Thanks for looking. Pickling sounds likely. It also turns out it's failing on the first object that hits the worker.

**Comment by mahmoud** [comment]
Pickle issue reproduced and ostensibly addressed in the commit above. T is pretty special, it definitely needs a smidge of special handling. Kurt and I pickle so little that it slipped our minds, but really who pickles directly these days? It's always these underlying libraries that put us in a sour state ;P

**Comment by mahmoud** [comment]
To circle back, the fix for this was released v18.3.0 last night. Full changelog here: https://github.com/mahmoud/glom/blob/master/CHANGELOG.md

Thanks again for the report!
